includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php

```diff
@@ -0,0 +1,167 @@
+<?php
+/**
+ * REST API Marketplace suggestions API
+ *
+ * Handles requests for marketplace suggestions data & rendering
+ * templates for suggestion DOM content.
+ *
+ * @package WooCommerce\Classes
+ * @since   3.6.0
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * REST API Marketplace suggestions API and related logic.
+ */
+class WC_Marketplace_Suggestions {
+
+	/**
+	 * Initialise.
+	 */
+	public static function init() {
+		if ( ! self::allow_suggestions() ) {
+			return;
+		}
+
+		// Register ajax api handlers.
+		add_action( 'wp_ajax_woocommerce_marketplace_suggestions', array( __CLASS__, 'get_suggestion_json_data_handler' ) );
+		add_action( 'wp_ajax_woocommerce_add_dismissed_marketplace_suggestion', array( __CLASS__, 'post_add_dismissed_suggestion_handler' ) );
+
+		// Register hooks for rendering suggestions container markup.
+		add_action( 'wc_marketplace_suggestions_products_empty_state', array( __CLASS__, 'render_products_list_empty_state' ) );
+	}
+
+	/**
+	 * Suggestion data GET handler.
+	 */
+	public static function get_suggestion_json_data_handler() {
+		$suggestion_data = self::get_suggestions_api_data();
```

Comment:
Instead of calling API when a ajax requests for suggestions are made, why not have a backrground process that runs every x amount of days and then cache the data in the DB and rather look for that option instead? This should solve the issue of failing API requests and visits to certain pages retriggering the requests.

WC_Queue is your friend https://github.com/woocommerce/woocommerce/wiki/WC_Queue---WooCommerce-Worker-Queue

Reply:
This sounds perfect! I had trouble getting it to work though – perhaps I'm missing something?

Here's a minimal example (based on the docs):

```php
function my_job_function( $param1, $param2 ) {
    // This is the code that will execute when the job runs.
    error_log( 'queue job running' );

    error_log( 'Param1: ' . $param1 );
    error_log( 'Param2: ' . $param2 );
}
add_action( 'my_job_action', 'my_job_queue', 10, 2 );
error_log( 'registered job' );
WC()->queue()->add( 'my_job_action', array(
	'param1' => 'test',
	'param2' => 'code',
) );
```

When I run this, I get `registered job` in the log, but I haven't managed to confirm the job actually runs. 

I've commited the work-in-progress code to another branch in case we can salvage it. @kovshenin  any ideas?

https://github.com/haszari/woocommerce/commit/4940aad6eaf8d9bb87eda6d271223f14be8483f1

Reply:
You might want `add_action( 'my_job_action', 'my_job_function', 10, 2 );` instead? (replace `my_job_queue`)

Reply:
@willgorham dang, good spotting! However, still no luck.

(Edited the code example above after a retest.)